rust/nft-packs/src/processor/add_card_to_pack.rs

```diff
@@ -0,0 +1,133 @@
+    utils::{assert_derivation, assert_initialized},
+};
+
+/// Process AddCardToPack instruction
+pub fn add_card_to_pack(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    args: AddCardToPackArgs,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let pack_set_info = next_account_info(account_info_iter)?;
+    let pack_card_info = next_account_info(account_info_iter)?;
+    let authority_info = next_account_info(account_info_iter)?;
+    let master_edition_info = next_account_info(account_info_iter)?;
+    let master_metadata_info = next_account_info(account_info_iter)?;
+    let mint_info = next_account_info(account_info_iter)?;
+    let source_info = next_account_info(account_info_iter)?;
+    let token_account_info = next_account_info(account_info_iter)?;
+    let rent_info = next_account_info(account_info_iter)?;
+    let rent = &Rent::from_account_info(rent_info)?;
+
+    assert_signer(authority_info)?;
+    assert_rent_exempt(rent, pack_card_info)?;
+    assert_owned_by(pack_set_info, program_id)?;
+    assert_owned_by(master_edition_info, &spl_token_metadata::id())?;
+    assert_owned_by(master_metadata_info, &spl_token_metadata::id())?;
+
+    let AddCardToPackArgs {
+        max_supply,
+        probability_type,
+        probability,
+        index,
+    } = args;
+
+    let mut pack_set = PackSet::unpack(&pack_set_info.data.borrow_mut())?;
+    let mut pack_card = PackCard::unpack_unchecked(&pack_card_info.data.borrow_mut())?;
+    assert_uninitialized(&pack_card)?;
+
+    let (pack_card_pubkey, bump_seed) =
+        find_pack_card_program_address(program_id, pack_set_info.key, index);
+    assert_account_key(pack_card_info, &pack_card_pubkey)?;
+
+    let signers_seeds = &[
+        &pack_set_info.key.to_bytes()[..32],
+        "card".as_bytes(),
+        &args.index.to_be_bytes(),
+        &[bump_seed],
+    ];
+
+    create_account::<PackCard>(
+        program_id,
+        authority_info.clone(),
+        pack_card_info.clone(),
+        &[signers_seeds],
+        rent,
+    )?;
+
+    let token_metadata_program_id = spl_token_metadata::id();
+
+    // Check for v2
+    let _master_edition = MasterEditionV2::from_account_info(master_edition_info)?;
+
+    let master_metadata = Metadata::from_account_info(master_metadata_info)?;
+    assert_account_key(mint_info, &master_metadata.mint)?;
+    assert_derivation(
+        &token_metadata_program_id,
+        master_edition_info,
+        &[
+            PREFIX.as_bytes(),
+            token_metadata_program_id.as_ref(),
+            master_metadata.mint.as_ref(),
+            EDITION.as_bytes(),
+        ],
+    )?;
+
+    let source: Account = assert_initialized(source_info)?;
+    if source.mint != master_metadata.mint {
+        return Err(MetadataError::MintMismatch.into());
+    }
+
+    // Initialize token account
+    spl_initialize_account(
+        token_account_info.clone(),
+        mint_info.clone(),
+        authority_info.clone(),
```

Comment:
In initializing token account we should set program acc as a authority, not pack set authority